src/adaptors/throttle.rs

```diff
@@ -149,48 +150,34 @@ impl Default for Limits {
 pub struct Throttle<B> {
     bot: B,
     // Sender<Never> is used to pass the signal to unlock by closing the channel.
-    queue: mpsc::Sender<(Id, Sender<Never>)>,
+    queue: mpsc::Sender<(FastChatId, Sender<Never>)>,
 }
 
-async fn worker(limits: Limits, mut queue_rx: mpsc::Receiver<(Id, Sender<Never>)>) {
-    // +- Same idea as in `Throttle::new`
-    let cap = limits.overall_s + (limits.overall_s / 4);
+type RequestsSent = u32;
+
+// I wish there was special data structure for history which removed the
+// need in 2 hashmaps
+// (waffle)
+#[derive(Default)]
+struct RequestsSentToChats {
+    per_min: HashMap<FastChatId, RequestsSent>,
+    per_sec: HashMap<FastChatId, RequestsSent>,
+}
+
+async fn worker(limits: Limits, mut rx: mpsc::Receiver<(FastChatId, Sender<Never>)>) {
     // FIXME(waffle): Make an research about data structures for this queue.
     //                Currently this is O(n) removing (n = number of elements
     //                stayed), amortized O(1) push (vec+vecrem).
-    let mut queue: Vec<(Id, Sender<Never>)> = Vec::with_capacity(cap as usize);
-
-    // I wish there was special data structure for history which removed the
-    // need in 2 hashmaps
-    // (waffle)
-    let mut history: VecDeque<(Id, Instant)> = VecDeque::new();
-    // hchats[chat] = history.iter().filter(|(c, _)| c == chat).count()
-    let mut hchats: HashMap<Id, u32> = HashMap::new();
-    let mut hchats_s = HashMap::new();
-
-    // set to true when `queue_rx` is closed
-    let mut close = false;
-
-    while !close || !queue.is_empty() {
-        // If there are no pending requests we are just waiting
-        if queue.is_empty() {
-            match queue_rx.recv().await {
-                Some(req) => queue.push(req),
-                None => close = true,
-            }
-        }
+    let mut queue: Vec<(FastChatId, Sender<Never>)> =
+        Vec::with_capacity(limits.messages_per_sec_overall as usize);
 
-        // update local queue with latest requests
-        loop {
-            // FIXME(waffle): https://github.com/tokio-rs/tokio/issues/3350
-            match queue_rx.recv().now_or_never() {
-                Some(Some(req)) => queue.push(req),
-                // There are no items in queue
-                None => break,
-                // The queue was closed
-                Some(None) => close = true,
-            }
-        }
+    let mut when_requests_were_sent: VecDeque<(FastChatId, Instant)> = VecDeque::new();
+    let mut requests_sent_to_chats = RequestsSentToChats::default();
```

Comment:
IMO, this naming is **too** verbose. I think identifiers this long are not helping to read code but instead are overwhelming the reader.

I would better use `history` and `sent_requests`. Some points in addition to the length:
- `when_requests_were_sent` doesn't suggest that the deque is sorted by time
- all requests are sent to chats in some sense...